### PR TITLE
refactor: Refactor error in DataType

### PR DESCRIPTION
**Pull request by XuQianJin-Stars**
link: https://github.com/apache/paimon-rust/issues/32

### CODE REVIEW

File: crates/paimon/src/error.rs [hunk 1]
```diff
@@ -38,3 +38,12 @@ pub enum Error {
         source: opendal::Error,
     },
 }
+
+#[derive(Debug, Snafu)]
+pub enum DataTypeError {
```
Comment (Xuanwo) [inline crates/paimon/src/error.rs:43]: Hi, we usually expose only one error in each crate to keep them user-friendly. I will suggest to add `DataTypeInvalid` as a new enum in `Error` directly. In this way, we can use `paimon::Error` everywhere.

File: crates/paimon/src/spec/types.rs [hunk 2]
```diff
@@ -320,13 +323,16 @@ impl CharType {
...
-    pub fn new(length: usize) -> Result<Self, &'static str> {
+    pub fn new(length: usize) -> Result<Self, Error> {
         Self::with_nullable(true, length)
     }
 
-    pub fn with_nullable(nullable: bool, length: usize) -> Result<Self, &'static str> {
+    pub fn with_nullable(nullable: bool, length: usize) -> Result<Self, Error> {
         if !(Self::MIN_LENGTH..=Self::MAX_LENGTH).contains(&length) {
-            return Err("Character string length must be between 1 and 255 (both inclusive).");
+            return DataTypeInvalidSnafu {
```
Comment (Aitozi) [inline crates/paimon/src/spec/types.rs:332]: Can we simplify this by `ensure!(length < Self::MIN_LENGTH, DataTypeInvalidSnafu {message: "Binary string length must be at least 1."});` ?

The same as below

**Review by Aitozi — COMMENTED** [review]
Left a minor comment.

File: crates/paimon/src/spec/types.rs (hunk 2 above)
Comment (XuQianJin-Stars) [reply crates/paimon/src/spec/types.rs:332]: > Can we simplify this by `ensure!(length < Self::MIN_LENGTH, DataTypeInvalidSnafu {message: "Binary string length must be at least 1."});` ?
> 
> The same as below

This simplifies things a bit, but the readability of the code is reduced.

Comment (Aitozi) [reply crates/paimon/src/spec/types.rs:332]: Ok, make sense

**Review by Aitozi — APPROVED** [review]
LGTM

**Review by Xuanwo — APPROVED** [review]
THanks

**Pull request merged**